okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/UrlSelectorImpl.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting3.okhttp;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import java.util.Collection;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import okhttp3.HttpUrl;
+
+final class UrlSelectorImpl implements UrlSelector {
+
+    private final ImmutableList<HttpUrl> baseUrls;
+    private final AtomicInteger currentUrl;
+
+    private UrlSelectorImpl(ImmutableList<HttpUrl> baseUrls) {
+        this.baseUrls = baseUrls;
+        this.currentUrl = new AtomicInteger(0);
+    }
+
+    static UrlSelectorImpl create(Collection<String> baseUrls) {
+        ImmutableSet.Builder<HttpUrl> canonicalUrls = ImmutableSet.builder();
+        baseUrls.forEach(url -> {
+            HttpUrl httpUrl = HttpUrl.parse(url);
+            Preconditions.checkArgument(httpUrl != null, "Not a valid URL: %s", url);
+            HttpUrl canonicalUrl = canonicalize(httpUrl);
+            Preconditions.checkArgument(canonicalUrl.equals(httpUrl),
+                    "Base URLs must be 'canonical' and consist of schema, host, port, and path only: %s", url);
+            canonicalUrls.add(canonicalUrl);
+        });
+        return new UrlSelectorImpl(ImmutableList.copyOf(canonicalUrls.build()));
+    }
+
+    @Override
+    public Optional<HttpUrl> redirectTo(HttpUrl current, String redirectBaseUrl) {
+        return redirectTo(current, HttpUrl.parse(redirectBaseUrl));
+    }
+
+    private Optional<HttpUrl> redirectTo(HttpUrl current, HttpUrl redirectBaseUrl) {
+        return baseUrlFor(redirectBaseUrl).flatMap(baseUrl -> {
```

Comment:
Can we simplify a bunch of this if we require all urls to have the same base path at construction time?

Reply:
I think you still need to do the prefix matching, so it probably doesn't make a big difference?

Reply:
Is it that where we do all this work we no longer know the distinction between the application's requested path and the URL we're adding it to?

Reply:
yeah.

Reply:
we construct the okhttp interceptor with the discovery-consumed service URLs. we then intercept a feign/retrofit - generated URL that contains richer paths, parameters, query parameters, etc. redirecting here means replacing its base path by another one.